modules/tinymce/src/core/main/ts/lookup/UserLookup.ts

```diff
@@ -0,0 +1,214 @@
+   * @return {Promise<User>[]} A promise that resolves to an array of users and information about them.
+   */
+  fetchUsers: (userIds: UserId[]) => Promise<User>[];
+}
+
+const userSchema = StructureSchema.objOf([
+  FieldSchema.required('id'),
+  FieldSchema.optionString('name'),
+  FieldSchema.optionString('avatar'),
+  FieldSchema.optionString('description'),
+  FieldSchema.option('custom')
+]);
+
+const objectCat = <T extends object>(
+  obj: { [K in keyof T]: Optional<T[K]> }
+): Partial<T> => {
+  const result = {} as Partial<T>;
+
+  Obj.each(obj, (value, key) => {
+    value.each((v) => {
+      result[key as keyof T] = v;
+    });
+  });
+
+  return result;
+};
+
+const validateResponse = (items: unknown): User[] => {
+  if (!Array.isArray(items)) {
+    throw new Error('fetch_users must return an array');
+  }
+
+  const results = Arr.map(items, (item) =>
+    StructureSchema.asRaw<ValidatedUser>('Invalid user object', userSchema, item)
+  );
+
+  const { errors, values } = Results.partition(results);
+
+  if (errors.length > 0) {
+    const formattedErrors = Arr.map(errors, (error, idx) =>
+      `User at index ${idx}: ${StructureSchema.formatError(error)}`
+    );
+
+    // eslint-disable-next-line no-console
+    console.warn('User validation errors:\n' + formattedErrors.join('\n'));
+  }
+
+  return Arr.map(values, (user) => {
+    const { id, ...rest } = user;
+
+    return {
+      id,
+      ...objectCat(rest),
+    };
+  });
+};
+
+const UserLookup = (editor: Editor): UserLookup => {
+  const userCache = new Map<UserId, Promise<User>>();
+  const pendingResolvers = new Map<UserId, {
+    resolve: (user: User) => void;
+    reject: (error: unknown) => void;
+  }>();
+
+  const lookup = (userId: UserId) =>
+    Optional.from(userCache.get(userId));
+
+  const store = (user: Promise<User>, userId: UserId) => {
+    userCache.set(userId, user);
+  };
+
+  const finallyReject = (userId: UserId, error: Error) => {
+    return Optional.from(pendingResolvers.get(userId)).each(({ reject }) => {
+      reject(error);
+      pendingResolvers.delete(userId);
+    });
+  };
+
+  const fetchUsers = (userIds: UserId[]): Promise<User>[] => {
+    if (!Array.isArray(userIds)) {
+      return [];
+    }
+
+    const { fail: uncachedIds } = Arr.partition(userIds, (userId) =>
+      lookup(userId).isSome()
+    );
+
+    Arr.each(uncachedIds, (userId) => {
+      const newPromise = new Promise<User>((resolve, reject) => {
+        pendingResolvers.set(userId, { resolve, reject });
+      });
+      store(newPromise, userId);
+    });
+
+    if (uncachedIds.length > 0) {
+      Optional.from(Options.getFetchUsers(editor)).fold(
+        () => {
+          Arr.each(uncachedIds, (userId) => {
+            const pending = pendingResolvers.get(userId);
+            if (pending) {
+              pending.resolve({ id: userId });
+              pendingResolvers.delete(userId);
+            }
+          });
+        },
```

Comment:
```suggestion
        () => Arr.each(uncachedIds, finallyResolve),
```
So it can be used here

Reply:
Nice! I ended up going with `(user) => finallyResolve(user)` because `Arr.each` would pass an index number as a second parameter which isn't what `finallyResolve` expects leading to a type mismatch so we have to just pass it what it needs but still does the trick.